pkg/controller/deployment_rolling.go

```diff
@@ -82,11 +89,28 @@ func (dc *controller) rolloutRolling(d *v1alpha1.MachineDeployment, isList []*v1
 	}
 
 	if MachineDeploymentComplete(d, &d.Status) {
+		// Check if any of the machine under this MachineDeployment contains the by-mcm annotation, and
+		// remove the original autoscaler-annotion only after.
+		err := dc.removeAutoscalerAnnotationsIfRequired(allISs, clusterAutoscalerScaleDownAnnotations)
+		if err != nil {
+			return err
+		}
 		if err := dc.cleanupMachineDeployment(oldISs, d); err != nil {
 			return err
 		}
 	}
 
+	// Check if all the oldMachineSets are set to zero, if not there is likely be a roll-out ongoing.
+	if len(oldISs) > 0 && !MachineDeploymentComplete(d, &d.Status) && !dc.machineSetsScaledToZero(oldISs) {
```

Comment:
This logic adds the annotations on all the node-objects immediately the roll-out starts, seems that's the best MCM could in this case.

Reply:
> annotations on all the node-objects immediately the roll-out starts, seems that's the best MCM could in this case.

This is true. But that is not what the code in this PR does. The code annotates the nodes before the deployment status is synced. Machine roll-out starts already earlier with the [tainting of the old nodes](https://github.com/gardener/machine-controller-manager/pull/496/files#diff-d60d538aae4eb1c57f7cd31e87804382R60-R66) and the scaling of [new](https://github.com/gardener/machine-controller-manager/pull/496/files#diff-d60d538aae4eb1c57f7cd31e87804382R72) and [old](https://github.com/gardener/machine-controller-manager/pull/496/files#diff-d60d538aae4eb1c57f7cd31e87804382L75) `machinesets`. Or am I missing something?

Reply:
Ah, sorry, I mistook the taint by the ones put by the autoscaler, but I guess you meant the `PreferNoSchedule` taint from MCM. 
Can you please elaborate a bit around, how `PreferNoSchedule` taint and `ScaleDown` annotation can impact each other's function?

Reply:
> Recreate strategy simply deletes all the existing machines and creates the new ones altogether. That is already super disruptive, not sure how autoscaler could harm it any further.

What if CA scales down/up while the new machines are still joining? Isn't it better to cover all bases?

Reply:
>Ah, sorry, I mistook the taint by the ones put by the autoscaler, but I guess you meant the PreferNoSchedule taint from MCM.

Yes, I was talking about the `PreferNoSchedule` taints. But not only that. I was also talking about `reconcileNewMachineSet` and `reconcileOldMachineSets` which also happen before adding the annotation in the current PR. Not only that, the code path of `reconcileNewMachineSet` and `reconcileOldMachineSets` might even [return](https://github.com/gardener/machine-controller-manager/blob/3def9452ec9b1796e5f2171616abc31a0ce7f84e/pkg/controller/deployment_rolling.go#L67) [from](https://github.com/gardener/machine-controller-manager/blob/master/pkg/controller/deployment_rolling.go#L71) reconciliation [without](https://github.com/gardener/machine-controller-manager/blob/3def9452ec9b1796e5f2171616abc31a0ce7f84e/pkg/controller/deployment_rolling.go#L77) [adding](https://github.com/gardener/machine-controller-manager/blob/3def9452ec9b1796e5f2171616abc31a0ce7f84e/pkg/controller/deployment_rolling.go#L81) the annotation. Isn't it better to annotate the nodes before all this?

Reply:
Hm, that's a good point. Though it's not possible to simply move the [logic](https://github.com/gardener/machine-controller-manager/blob/44d70a27e902b52d98d44da7717bc4ad471d099d/pkg/controller/deployment_rolling.go#L104-L112) up before tainting, else the `MachineDeploymentComplete` will return `true`, and MCM may not end up annotating at all.

But I do see the point that we don't want to see the unannotated nodes only because the reconciliation failed in-between for unknown reasons. I'll add the logic to annotate only the `OldMachineSets` near the taint-mechanism.

Reply:
Can you please check if this change: https://github.com/gardener/machine-controller-manager/blob/b99e8a8371e2bb1f7de25461b26970e4b3e99ed4/pkg/controller/deployment_rolling.go#L71-L78  , serves the purpose?
- I'd expect, at the end when the MachineDeployment is complete [here](https://github.com/gardener/machine-controller-manager/blob/b99e8a8371e2bb1f7de25461b26970e4b3e99ed4/pkg/controller/deployment_rolling.go#L104), the annotations will be removed altogether.

Reply:
Thanks! That looks better. But now the check on `err` from `taintNodesBackingMachineSets` could potentially overwritten before checking.

Basically, the following has to be immediately after the taint call.

https://github.com/gardener/machine-controller-manager/blob/b99e8a8371e2bb1f7de25461b26970e4b3e99ed4/pkg/controller/deployment_rolling.go#L80-L82

Reply:
BTW, with this change I am not sure if the later annotate call is required at all.

https://github.com/gardener/machine-controller-manager/blob/b99e8a8371e2bb1f7de25461b26970e4b3e99ed4/pkg/controller/deployment_rolling.go#L117-L125

Reply:
Yes, that's true. I have also realized the same and experimenting a bit to make sure we don't lose a corner case. I'll most likely remove the annotation-logic put later.

Reply:
Done, removed.